src/backend_memory.cc

```diff
@@ -26,53 +26,141 @@
 
 #include "triton/backend/backend_memory.h"
 
+#include <map>
 #include "triton/backend/backend_common.h"
 
 namespace triton { namespace backend {
 
 TRITONSERVER_Error*
 BackendMemory::Create(
-    TRITONBACKEND_MemoryManager* manager,
-    const TRITONSERVER_MemoryType memory_type, const int64_t memory_type_id,
-    const size_t byte_size, BackendMemory** mem)
+    TRITONBACKEND_MemoryManager* manager, const AllocationType alloc_type,
+    const int64_t memory_type_id, const size_t byte_size, BackendMemory** mem)
 {
   *mem = nullptr;
 
+  TRITONSERVER_MemoryType memory_type = TRITONSERVER_MEMORY_CPU;
   void* ptr = nullptr;
-  RETURN_IF_ERROR(TRITONBACKEND_MemoryManagerAllocate(
-      manager, &ptr, memory_type, memory_type_id, byte_size));
+  switch (alloc_type) {
+    case AllocationType::CPU: {
+      memory_type = TRITONSERVER_MEMORY_CPU;
+      ptr = malloc(byte_size);
+      RETURN_ERROR_IF_TRUE(
+          ptr == nullptr, TRITONSERVER_ERROR_UNAVAILABLE,
+          std::string("CPU memory allocation failed"));
+      break;
+    }
+
+    case AllocationType::CPU_PINNED: {
+      memory_type = TRITONSERVER_MEMORY_CPU_PINNED;
+
+#ifdef TRITON_ENABLE_GPU
+      RETURN_IF_CUDA_ERROR(
+          cudaHostAlloc(&ptr, byte_size, cudaHostAllocPortable),
+          TRITONSERVER_ERROR_UNAVAILABLE,
+          std::string("failed to allocate pinned system memory"));
+#else
+      return TRITONSERVER_ErrorNew(
+          TRITONSERVER_ERROR_UNSUPPORTED,
+          "pinned-memory allocation not supported");
+#endif  // TRITON_ENABLE_GPU
+      break;
+    }
+
+    case AllocationType::GPU: {
+      memory_type = TRITONSERVER_MEMORY_GPU;
+
+#ifdef TRITON_ENABLE_GPU
+      int current_device;
+      RETURN_IF_CUDA_ERROR(
+          cudaGetDevice(&current_device), TRITONSERVER_ERROR_INTERNAL,
+          std::string("failed to get device"));
+      bool overridden = (current_device != memory_type_id);
+      if (overridden) {
+        RETURN_IF_CUDA_ERROR(
+            cudaSetDevice(memory_type_id), TRITONSERVER_ERROR_INTERNAL,
+            std::string("failed to set device"));
+      }
+
+      auto err = cudaMalloc(&ptr, byte_size);
+
+      if (overridden) {
+        cudaSetDevice(current_device);
+      }
+
+      RETURN_ERROR_IF_FALSE(
+          err == cudaSuccess, TRITONSERVER_ERROR_UNAVAILABLE,
+          std::string("failed to allocate GPU memory: ") +
+              cudaGetErrorString(err));
+#else
+      return TRITONSERVER_ErrorNew(
+          TRITONSERVER_ERROR_UNSUPPORTED, "GPU allocation not supported");
+#endif  // TRITON_ENABLE_GPU
+      break;
+    }
+
+    case AllocationType::CPU_PINNED_POOL: {
+      memory_type = TRITONSERVER_MEMORY_CPU_PINNED;
+      RETURN_IF_ERROR(TRITONBACKEND_MemoryManagerAllocate(
+          manager, &ptr, TRITONSERVER_MEMORY_CPU_PINNED, memory_type_id,
+          byte_size));
+      break;
+    }
+
+    case AllocationType::GPU_POOL: {
+      memory_type = TRITONSERVER_MEMORY_GPU;
```

Comment:
wrap in ifdef/else/endif TRITON_ENABLE_GPU

Reply:
This is taken care of inside of TRITONBACKEND_MemoryManagerAllocate